lib/puppet/pops/evaluator/closure.rb

```diff
@@ -122,6 +122,18 @@ def initialize(name, evaluator, model, scope)
     def closure_name
       @name
     end
+
+    # The assigned enclosing scope, or global scope if enclosing scope was initialized to nil
+    #
+    def enclosing_scope
+      # Named closures are typically used for puppet functions and they cannot be defined
+      # in an enclosing scope as they are cashed and reused. They need to bind to the
+      # global scope at time of use rather at time of definition.
+      # Unnamed closures are always a runtime construct, they are never bound by a loader
+      # and are thus garbage collected at end of a compilation.
+      #
+      super || Puppet.lookup(:global_scope) { {} }
+    end
```

Comment:
There is a lingering reference to the attribute @enclosing_scope at line 248. That's whats causing the problems in the unit tests.